mixnet/node/src/lib.rs

```diff
@@ -104,178 +101,229 @@ impl MixnetNode {
-        client_tx: &mpsc::Sender<Body>,
-        body: Body,
-    ) -> Result<()> {
+    async fn forward_body_to_client_notifier(&self, body: Body) -> Result<()> {
         // TODO: Decrypt the final payload using the private key, if it's encrypted
 
         // Do not wait when the channel is full or no receiver exists
-        client_tx.try_send(body)?;
+        self.client_tx.try_send(body)?;
         Ok(())
     }
 
     async fn forward_packet_to_next_hop(
-        pool: &ConnectionPool,
-        max_retries: usize,
-        retry_delay: Duration,
-        packet: Box<SphinxPacket>,
+        &self,
+        packet: Message,
         next_node_addr: NodeAddressBytes,
         delay: Delay,
     ) -> Result<()> {
         tracing::debug!("Delaying the packet for {delay:?}");
         tokio::time::sleep(delay.to_duration()).await;
 
-        Self::forward(
-            pool,
-            max_retries,
-            retry_delay,
-            Body::new_sphinx(packet),
-            NymNodeRoutingAddress::try_from(next_node_addr)?,
-        )
-        .await
+        self.forward(packet, NymNodeRoutingAddress::try_from(next_node_addr)?)
+            .await
     }
 
     async fn forward_payload_to_destination(
-        pool: &ConnectionPool,
-        max_retries: usize,
-        retry_delay: Duration,
-        payload: Payload,
+        &self,
+        payload: Message,
         destination_addr: DestinationAddressBytes,
     ) -> Result<()> {
         tracing::debug!("Forwarding final payload to destination mixnode");
 
-        Self::forward(
-            pool,
-            max_retries,
-            retry_delay,
-            Body::new_final_payload(payload),
+        self.forward(
+            payload,
             NymNodeRoutingAddress::try_from_bytes(&destination_addr.as_bytes())?,
         )
         .await
     }
 
-    async fn forward(
-        pool: &ConnectionPool,
-        max_retries: usize,
-        retry_delay: Duration,
-        body: Body,
-        to: NymNodeRoutingAddress,
-    ) -> Result<()> {
+    async fn forward(&self, msg: Message, to: NymNodeRoutingAddress) -> Result<()> {
         let addr = SocketAddr::from(to);
-        let arc_socket = pool.get_or_init(&addr).await?;
-
-        if let Err(e) = {
-            let mut socket = arc_socket.lock().await;
-            body.write(&mut *socket).await
-        } {
-            tracing::error!("Failed to forward packet to {addr} with error: {e}. Retrying...");
-            return mixnet_protocol::retry_backoff(
-                addr,
-                max_retries,
-                retry_delay,
-                body,
-                arc_socket,
-            )
-            .await
-            .map_err(Into::into);
-        }
+
+        let body = if let Some(handle) = self.message_pool.get(&addr) {
+            if let Err(err) = handle.send(msg) {
+                err.0
+            } else {
+                // If succesfull send message to msg writting thread,
```

Comment:
is it a writing thread or a writing task?

Reply:
It should be sent to the writing thread here, if we can get a handle, which means that the writing thread for such addr is living, and put msg to the channel, in the future, it will be sent to the remote.